### PR TITLE
#1288 installer config and tests

### PR DESCRIPTION
**Pull request by DonJayamanne**
Support for workspace in the following areas:
- Installer (installation of python packages such as linters, test runners)  
- Config settings (getting resource specific config settings)  
- Changes to linters and formatters to make use of the new config settings  
- Unit tests for multi root workspace

### CODE REVIEW

File: src/client/common/installer.ts [hunk 1]
```diff
@@ -189,46 +199,50 @@ export class Installer implements vscode.Disposable {
...
+            case useOtherFormatter: {
+                return this.updateSetting('formatting.provider', alternateFormatter, resource)
+                    .then(() => InstallerResponse.Installed);
+            }
+            case dontShowAgain: {
+                const pythonConfig = workspace.getConfiguration('python');
+                const features = pythonConfig.get('disablePromptForFeatures', [] as string[]);
+                features.push(productName);
+                return pythonConfig.update('disablePromptForFeatures', features, true).then(() => InstallerResponse.Ignore);
+            }
```
Comment (brettcannon) [inline src/client/common/installer.ts]: Does TypeScript make sure that all possible case statements are covered?

File: src/test/common/configSettings.multiroot.test.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    test('Enabling/Disabling Pylint in root should be reflected in config settings', async () => {
+        const workspaceUri = Uri.file(path.join(multirootPath, 'workspace1'));
+        await testLinterSetting(workspaceUri, ConfigurationTarget.Workspace, 'pylintEnabled', true);
+        await testLinterSetting(workspaceUri, ConfigurationTarget.Workspace, 'pylintEnabled', false);
+    });
+
+    test('Enabling/Disabling Pylint in root and workspace should be reflected in config settings', async () => {
+        const workspaceUri = Uri.file(path.join(multirootPath, 'workspace1'));
+        const workspaceFolder = Uri.file(path.join(multirootPath, 'workspace1'));
```
Comment (brettcannon) [inline src/test/common/configSettings.multiroot.test.ts]: How is `workspaceFolder` different from `workspaceUri`? The code looks the same.

File: src/test/multiRootWkspc/disableLinters/file.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+"""pylint option block-disable"""
```
Comment (brettcannon) [inline src/test/multiRootWkspc/disableLinters/file.py]: Probably should either (a) make this work under Python 3, or (b) add a comment that this code is actually never run so it shouldn't effect if Python 2 or 3 is used. (Same for other test files).

File: src/test/common/configSettings.multiroot.test.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/test/common/configSettings.multiroot.test.ts]: wtf

File: src/test/multiRootWkspc/disableLinters/file.py (hunk 3 above)
Comment (DonJayamanne) [reply src/test/multiRootWkspc/disableLinters/file.py]: Will do

File: src/client/common/installer.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/common/installer.ts]: No it doesn't, will add one

File: src/test/multiRootWkspc/disableLinters/file.py (hunk 3 above)
Comment (DonJayamanne) [reply src/test/multiRootWkspc/disableLinters/file.py]: Fixed

File: src/test/common/configSettings.multiroot.test.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/test/common/configSettings.multiroot.test.ts]: Fixed

File: src/client/common/installer.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/common/installer.ts]: Fixed

File: src/test/multiRootWkspc/disableLinters/file.py [hunk 4]
```diff
@@ -10,78 +10,78 @@ def __init__(self):
...
+        print (self.blip)
 
     def meth5(self):
         """test IF sub-block re-enabling"""
         # pylint: disable=no-member
         # no error
-        print self.bla
-        if self.blop:
+        print (self.bla)
+        if (self.blop):
```
Comment (brettcannon) [inline src/test/multiRootWkspc/disableLinters/file.py]: Got overly zealous with the parentheses. 😉

File: src/test/multiRootWkspc/parent/child/file.py [hunk 5]
```diff
@@ -10,78 +10,78 @@ def __init__(self):
...
+        print (self.blip)
 
     def meth5(self):
         """test IF sub-block re-enabling"""
         # pylint: disable=no-member
         # no error
-        print self.bla
-        if self.blop:
+        print (self.bla)
+        if (self.blop):
```
Comment (brettcannon) [inline src/test/multiRootWkspc/parent/child/file.py]: Still zealous 😉

File: src/test/multiRootWkspc/workspace1/file.py [hunk 6]
```diff
@@ -10,78 +10,78 @@ def __init__(self):
...
+        print (self.blip)
 
     def meth5(self):
         """test IF sub-block re-enabling"""
         # pylint: disable=no-member
         # no error
-        print self.bla
-        if self.blop:
+        print (self.bla)
+        if (self.blop):
```
Comment (brettcannon) [inline src/test/multiRootWkspc/workspace1/file.py]: You know what this is about 😄

File: src/test/multiRootWkspc/workspace3/file.py [hunk 7]
```diff
@@ -10,78 +10,78 @@ def __init__(self):
...
+        print (self.blip)
 
     def meth5(self):
         """test IF sub-block re-enabling"""
         # pylint: disable=no-member
         # no error
-        print self.bla
-        if self.blop:
+        print (self.bla)
+        if (self.blop):
```
Comment (brettcannon) [inline src/test/multiRootWkspc/workspace3/file.py]: Hello again

File: src/test/multiRootWkspc/workspace1/file.py (hunk 6 above)
Comment (DonJayamanne) [reply src/test/multiRootWkspc/workspace1/file.py]: argghh, welcome to python world..

**Pull request merged**